text/0000-declarative-shadow-dom-polyfill.md

```diff
@@ -0,0 +1,346 @@
+
+This approach is an attempt to get closer to the streaming (and lower CLS) inline script solution, while attempting to mitigate the parsing and compilation time required for repeated inline scripts. The major **disadvantages** here are:
+
+* This MutationObserver must be placed on `document.body`, and must observe all node additions to the document. This significantly slows down tree building, given the JS overhead for each element.
+* There is no "children finished parsing" mutation event. This means that the mutation observer in this polyfill must simply assume that by the time the callback is called, all of the children of the given `<template>` have already been parsed and appended to the tree. If the parser happens to yield in the middle of a `<template>` block, part of the children will be lost by this polyfill.
+
+```html
+<body>
+    <script type="text/javascript">
+        function supportsDeclarativeShadowDOM() {
+            return HTMLTemplateElement.prototype.hasOwnProperty("shadowRoot");
+        }
+        function attachShadowRoots(mutationsList) {
+            for (let mutation of mutationsList) {
+                for (let n of mutation.addedNodes) {
+                    if (n.tagName === 'TEMPLATE') {
+                        const mode = n.getAttribute('shadowroot');
+                        const shadowRoot = n.parentElement.attachShadow({mode});
+                        shadowRoot.appendChild(n.content);
+                        n.remove();
+                    }
+                }
+            }
+        }
+
+        if (!supportsDeclarativeShadowDOM()) {
+            (new MutationObserver(attachShadowRoots)).observe(document.body,
+                { childList: true, subtree: true, attributeFilter: [ "shadowroot" ], });
+        }
+    </script>
+    <div>
+        <template shadowroot="open"></template>
+    </div>
+</body>
+```
+
+### Performance comparison in Chrome, Safari, and Firefox
+
+These performance tests are using [tachometer](https://www.npmjs.com/package/tachometer), and compares the three different approach using 5k web components on the page. To eliminate forced style/layout, SSR-rendered web components are inside a `<div>` with `display:none` and `contain:strict` . The code used for this comparison can be found [here](https://github.com/jodarove/dec-shadow-dom-polyfill-spike#performance-tests).
+
+**CHROME**
+Despite Chrome natively supporting declarative shadow dom, we measure the speeds of the polyfills, because in the case of the inline script, the script element must be removed.
+
+<figure>
+  <img src="../assets/declarative-shadow-dom-polyfill/perf-chrome.png" alt="Performance comparison in Google Chrome">
+  <figcaption align = "center">
+    <b>Fig.1: Performance comparison in Google Chrome</b>
+  </figcaption>
+</figure>
+<br></br>
```

Comment:
I don't understand how it's possible for the single loop script to be slower than the inline script in Chrome. With the inline script, the browser needs to halt parsing after every `<template>` just to do a useless action (removing the `<script>`). Whereas with the single loop script, it does one useless action at the end (check `supportsDeclarativeShadowDOM`, then exit).